// @vueuse/motion/dist/nuxt.mjs
import { defineNuxtModule, createResolver, addPlugin, addImportsDir, resolveModule } from '@nuxt/kit';

const DEFAULTS = {};
const CONFIG_KEY = "motion";
const module = defineNuxtModule({
  meta: {
    name: "@vueuse/motion",
    configKey: CONFIG_KEY
  },
  defaults: DEFAULTS,
  setup(options, nuxt) {
    const { resolve } = createResolver(import.meta.url);
    const resolveRuntimeModule = (path) => resolveModule(path, { paths: resolve("./runtime") });
    nuxt.options.runtimeConfig.public.motion = options;
    addPlugin({
      src: resolveRuntimeModule("./templates/motion.js")
    });
    if (!nuxt.options.build.transpile)
      nuxt.options.build.transpile = [];
    const transpileList = ["defu", "@vueuse/motion", "@vueuse/shared", "@vueuse/core"];
    transpileList.forEach((pkgName) => {
      if (!nuxt.options.build.transpile.includes(pkgName))
        nuxt.options.build.transpile.push(pkgName);
    });
    if (!nuxt.options.alias)
      nuxt.options.alias = {};
    if (!nuxt.options.alias.tslib)
      nuxt.options.alias.tslib = "tslib/tslib.es6.js";
    addImportsDir(resolve("./composables/"));
  }
});

export { module as default };
