// @slidev/parser/dist/index.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});
















var _chunkVG22VYSBjs = require('./chunk-VG22VYSB.js');

















exports.detectFeatures = _chunkVG22VYSBjs.detectFeatures; exports.filterDisabled = _chunkVG22VYSBjs.filterDisabled; exports.getDefaultConfig = _chunkVG22VYSBjs.getDefaultConfig; exports.mergeFeatureFlags = _chunkVG22VYSBjs.mergeFeatureFlags; exports.parse = _chunkVG22VYSBjs.parse; exports.parseAspectRatio = _chunkVG22VYSBjs.parseAspectRatio; exports.parseRangeString = _chunkVG22VYSBjs.parseRangeString; exports.parseSlide = _chunkVG22VYSBjs.parseSlide; exports.prettify = _chunkVG22VYSBjs.prettify; exports.prettifySlide = _chunkVG22VYSBjs.prettifySlide; exports.resolveConfig = _chunkVG22VYSBjs.resolveConfig; exports.resolveFonts = _chunkVG22VYSBjs.resolveFonts; exports.scanMonacoModules = _chunkVG22VYSBjs.scanMonacoModules; exports.stringify = _chunkVG22VYSBjs.stringify; exports.stringifySlide = _chunkVG22VYSBjs.stringifySlide; exports.verifyConfig = _chunkVG22VYSBjs.verifyConfig;
